File: typescript/src/protocol.ts
import { createHash, randomUUID, sign, verify } from "node:crypto";

export const PROTOCOL_VERSION = "agent-action-assertion/v1";
export const ASSERTION_PREFIX = "agentctl-aav1";
export const MAX_TTL_SECONDS = 300;

export class ProtocolError extends Error {
  readonly code: string;

  constructor(code: string, message: string) {
    super(message);
    this.name = "ProtocolError";
    this.code = code;
  }
}

type JsonValue = null | boolean | string | number | JsonValue[] | { [key: string]: JsonValue };

function fail(code: string, message: string): never {
  throw new ProtocolError(code, message);
}

function validateJsonValue(value: unknown): asserts value is JsonValue {
  if (value === null || typeof value === "boolean" || typeof value === "string") return;
  if (typeof value === "number") {
    if (!Number.isFinite(value) || !Number.isSafeInteger(value)) {
      fail("INVALID_JSON_NUMBER", "only finite safe integers are allowed");
    }
    return;
  }
  if (Array.isArray(value)) {
    value.forEach(validateJsonValue);
    return;
  }
  if (typeof value === "object") {
    for (const [key, item] of Object.entries(value)) {
      if (typeof key !== "string") fail("INVALID_JSON_KEY", "object keys must be strings");
      validateJsonValue(item);
    }
    return;
  }
  fail("INVALID_JSON_VALUE", `unsupported JSON value type: ${typeof value}`);
}

function sortJson(value: JsonValue): JsonValue {
  if (Array.isArray(value)) return value.map(sortJson);
  if (value !== null && typeof value === "object") {
    return Object.fromEntries(
      Object.keys(value)
        .sort(compareCodePoints)
        .map((key) => [key, sortJson(value[key])]),
    );
  }
  return value;
}

export function canonicalJson(value: JsonValue): string {
  validateJsonValue(value);
  return JSON.stringify(sortJson(value));
}

export function canonicalJsonBytes(value: JsonValue): Buffer {
  return Buffer.from(canonicalJson(value), "utf8");
}

export function sha256Hex(value: Uint8Array): string {
  return createHash("sha256").update(value).digest("hex");
}

function compareCodePoints(left: string, right: string): number {
  const a = Array.from(left, (char) => char.codePointAt(0) ?? 0);
  const b = Array.from(right, (char) => char.codePointAt(0) ?? 0);
  for (let index = 0; index < Math.min(a.length, b.length); index += 1) {
    if (a[index] !== b[index]) return a[index] - b[index];
  }
  return a.length - b.length;
}

function validateString(value: unknown, field: string, pattern?: RegExp): string {
  if (typeof value !== "string" || value.length === 0 || value.trim() !== value) {
    fail("MALFORMED_ASSERTION", `${field} must be a non-empty trimmed string`);
  }
  if (/[\u0000-\u0020\u007f]/u.test(value)) {
    fail("MALFORMED_ASSERTION", `${field} contains control characters`);
  }
  if (pattern && !pattern.test(value)) fail("MALFORMED_ASSERTION", `${field} has an invalid format`);
  return value;
}

function validateInteger(value: unknown, field: string): number {
  if (typeof value !== "number" || !Number.isSafeInteger(value) || value < 0) {
    fail("MALFORMED_ASSERTION", `${field} must be a non-negative integer`);
  }
  return value;
}

function validatePercentEscapes(value: string, code: string): void {
  for (let index = 0; index < value.length; index += 1) {
    if (value[index] !== "%") continue;
    if (!/^[0-9A-Fa-f]{2}$/u.test(value.slice(index + 1, index + 3))) {
      fail(code, "invalid percent escape");
    }
    index += 2;
  }
}

function normalizePath(path: string): string {
  if (path.length === 0) path = "/";
  if (!path.startsWith("/")) fail("INVALID_PATH", "request path must start with '/'");
  if (/[\u0000-\u0020\u007f]/u.test(path)) fail("INVALID_PATH", "request path contains control characters or spaces");
  validatePercentEscapes(path, "INVALID_PATH");
  let output = "";
  for (let index = 0; index < path.length; index += 1) {
    if (path[index] !== "%") {
      output += path[index];
      continue;
    }
    const byte = Number.parseInt(path.slice(index + 1, index + 3), 16);
    const character = String.fromCharCode(byte);
    if ((byte < 128 && /[A-Za-z0-9]/u.test(character)) || "-._~".includes(character)) {
      output += character;
    } else {
      output += `%${byte.toString(16).toUpperCase().padStart(2, "0")}`;
    }
    index += 2;
  }
  if (output.split("/").some((segment) => segment === "." || segment === "..")) {
    fail("INVALID_PATH", "dot segments are not allowed");
  }
  return output;
}

function decodeQueryPart(value: string): string {
  try {
    return decodeURIComponent(value.replaceAll("+", " "));
  } catch (error) {
    fail("INVALID_QUERY", `query is not valid UTF-8: ${String(error)}`);
  }
}

function encodeRfc3986(value: string): string {
  return encodeURIComponent(value).replace(/[!'()*]/gu, (character) => `%${character.charCodeAt(0).toString(16).toUpperCase()}`);
}

function canonicalQuery(query: string): string {
  if (query.length === 0) return "";
  validatePercentEscapes(query, "INVALID_QUERY");
  const pairs = query.split("&").map((pair) => {
    const separator = pair.indexOf("=");
    const rawName = separator < 0 ? pair : pair.slice(0, separator);
    const rawValue = separator < 0 ? "" : pair.slice(separator + 1);
    return [decodeQueryPart(rawName), decodeQueryPart(rawValue)] as const;
  });
  pairs.sort((left, right) => compareCodePoints(left[0], right[0]) || compareCodePoints(left[1], right[1]));
  return pairs.map(([name, value]) => `${encodeRfc3986(name)}=${encodeRfc3986(value)}`).join("&");
}

function stripAbsoluteAuthority(target: string): string {
  const scheme = target.match(/^([A-Za-z][A-Za-z0-9+.-]*):\/\//u);
  if (!scheme) return target;
  if (!/^(http|https)$/iu.test(scheme[1])) fail("INVALID_PATH", "only HTTP and HTTPS URLs are supported");
  const remainder = target.slice(scheme[0].length);
  const authorityEnd = remainder.search(/[/?#]/u);
  const authority = authorityEnd < 0 ? remainder : remainder.slice(0, authorityEnd);
  if (authority.length === 0 || authority.includes("@")) fail("INVALID_PATH", "URL userinfo or empty authority is not allowed");
  return authorityEnd < 0 ? "/" : remainder.slice(authorityEnd);
}

export function canonicalRequestTarget(target: string): string {
  if (typeof target !== "string" || target.length === 0) fail("INVALID_PATH", "request target must be a non-empty string");
  if (target.includes("#")) fail("INVALID_PATH", "URL fragments are not allowed");
  const relative = stripAbsoluteAuthority(target);
  const separator = relative.indexOf("?");
  const path = separator < 0 ? relative : relative.slice(0, separator);
  const query = separator < 0 ? "" : relative.slice(separator + 1);
  const canonicalPath = normalizePath(path);
  const canonicalQueryValue = canonicalQuery(query);
  return canonicalQueryValue ? `${canonicalPath}?${canonicalQueryValue}` : canonicalPath;
}

function b64urlEncode(value: Uint8Array): string {
  return Buffer.from(value).toString("base64url");
}

function b64urlDecode(value: string): Buffer {
  if (!/^[A-Za-z0-9_-]+$/u.test(value)) fail("MALFORMED_ASSERTION", "invalid base64url segment");
  const decoded = Buffer.from(value, "base64url");
  if (decoded.toString("base64url") !== value) fail("MALFORMED_ASSERTION", "invalid base64url segment");
  return decoded;
}

export interface AssertionPayload {
  version: string;
  iss: string;
  sub: string;
  aud: string;
  environment: string;
  scope: string;
  iat: number;
  nbf: number;
  exp: number;
  jti: string;
  kid: string;
  http_method: string;
  canonical_path: string;
  content_type: string;
  body_sha256: string;
  request_id: string;
  principal_epoch: number;
  key_epoch: number;
  resource?: string;
  project?: string;
}

const PAYLOAD_FIELDS = new Set([
  "version", "iss", "sub", "aud", "environment", "scope", "iat", "nbf", "exp", "jti", "kid",
  "http_method", "canonical_path", "content_type", "body_sha256", "request_id", "principal_epoch",
  "key_epoch", "resource", "project",
]);

export function validatePayload(input: unknown): AssertionPayload {
  if (input === null || typeof input !== "object" || Array.isArray(input)) fail("MALFORMED_ASSERTION", "payload must be a JSON object");
  const value = input as Record<string, unknown>;
  const unknown = Object.keys(value).filter((key) => !PAYLOAD_FIELDS.has(key));
  if (unknown.length > 0) fail("MALFORMED_ASSERTION", `unknown assertion fields: ${unknown.sort(compareCodePoints).join(", ")}`);
  if (value.version !== PROTOCOL_VERSION) fail("WRONG_VERSION", "unsupported assertion version");
  for (const field of ["iss", "sub", "aud", "environment", "scope", "kid", "request_id"]) validateString(value[field], field);
  validateString(value.jti, "jti", /^[A-Za-z0-9._:-]{8,256}$/u);
  validateString(value.http_method, "http_method", /^[A-Z]+$/u);
  const path = validateString(value.canonical_path, "canonical_path");
  if (path !== canonicalRequestTarget(path)) fail("MALFORMED_ASSERTION", "canonical_path is not canonical");
  const contentType = value.content_type;
  if (typeof contentType !== "string" || contentType !== normalizeContentType(contentType)) fail("MALFORMED_ASSERTION", "content_type is not normalized");
  const digest = validateString(value.body_sha256, "body_sha256", /^[0-9a-f]{64}$/u);
  const iat = validateInteger(value.iat, "iat");
  const nbf = validateInteger(value.nbf, "nbf");
  const exp = validateInteger(value.exp, "exp");
  const principalEpoch = validateInteger(value.principal_epoch, "principal_epoch");
  const keyEpoch = validateInteger(value.key_epoch, "key_epoch");
  if (value.iss !== value.sub) fail("ISSUER_SUBJECT_MISMATCH", "V1 does not support delegated subjects");
  if (iat > nbf || exp <= nbf) fail("INVALID_TIME_WINDOW", "iat <= nbf < exp is required");
  if (exp - iat > MAX_TTL_SECONDS) fail("TTL_EXCEEDED", "assertion TTL exceeds the V1 maximum");
  for (const field of ["resource", "project"]) if (field in value) validateString(value[field], field);
  return {
    version: PROTOCOL_VERSION,
    iss: value.iss as string,
    sub: value.sub as string,
    aud: value.aud as string,
    environment: value.environment as string,
    scope: value.scope as string,
    iat,
    nbf,
    exp,
    jti: value.jti as string,
    kid: value.kid as string,
    http_method: value.http_method as string,
    canonical_path: path,
    content_type: contentType,
    body_sha256: digest,
    request_id: value.request_id as string,
    principal_epoch: principalEpoch,
    key_epoch: keyEpoch,
    ...(typeof value.resource === "string" ? { resource: value.resource } : {}),
    ...(typeof value.project === "string" ? { project: value.project } : {}),
  };
}

export function normalizeContentType(value: string | undefined): string {
  if (value === undefined) return "";
  if (typeof value !== "string") fail("INVALID_CONTENT_TYPE", "content type must be a string");
  const normalized = value.trim().toLowerCase();
  if (/[\u0000-\u001f\u007f]/u.test(normalized)) fail("INVALID_CONTENT_TYPE", "content type contains control characters");
  return normalized;
}

export interface ParsedAssertion {
  payload: AssertionPayload;
  payloadSegment: string;
  signature: Buffer;
  compact: string;
}

export function buildAssertion(options: {
  principalId: string;
  audience: string;
  environment: string;
  scope: string;
  httpMethod: string;
  target: string;
  body: Uint8Array;
  keyId: string;
  principalEpoch: number;
  keyEpoch: number;
  privateKey: import("node:crypto").KeyObject;
  contentType?: string;
  resource?: string;
  project?: string;
  requestId?: string;
  jti?: string;
  now?: number;
  ttlSeconds?: number;
}): string {
  const now = options.now ?? Math.floor(Date.now() / 1000);
  const ttl = options.ttlSeconds ?? MAX_TTL_SECONDS;
  if (!Number.isSafeInteger(now) || now < 0) fail("INVALID_TIME_WINDOW", "now must be a non-negative integer");
  if (!Number.isSafeInteger(ttl) || ttl <= 0 || ttl > MAX_TTL_SECONDS) fail("TTL_EXCEEDED", "ttl_seconds must be between 1 and 300");
  if (typeof options.httpMethod !== "string") fail("INVALID_METHOD", "HTTP method must be a string");
  const payload = validatePayload({
    aud: validateString(options.audience, "audience"),
    body_sha256: sha256Hex(options.body),
    canonical_path: canonicalRequestTarget(options.target),
    content_type: normalizeContentType(options.contentType),
    environment: validateString(options.environment, "environment"),
    exp: now + ttl,
    iat: now,
    iss: validateString(options.principalId, "principal_id"),
    jti: options.jti ?? randomUUID(),
    kid: validateString(options.keyId, "key_id"),
    nbf: now,
    http_method: options.httpMethod.toUpperCase(),
    principal_epoch: options.principalEpoch,
    request_id: options.requestId ?? randomUUID(),
    scope: validateString(options.scope, "scope"),
    sub: validateString(options.principalId, "principal_id"),
    version: PROTOCOL_VERSION,
    key_epoch: options.keyEpoch,
    ...(options.resource === undefined ? {} : { resource: options.resource }),
    ...(options.project === undefined ? {} : { project: options.project }),
  });
  const payloadSegment = b64urlEncode(canonicalJsonBytes(payload as unknown as JsonValue));
  const signingInput = Buffer.from(`${ASSERTION_PREFIX}.${payloadSegment}`, "ascii");
  return `${ASSERTION_PREFIX}.${payloadSegment}.${b64urlEncode(sign(null, signingInput, options.privateKey))}`;
}

export function parseAssertion(compact: string): ParsedAssertion {
  if (typeof compact !== "string") fail("MALFORMED_ASSERTION", "assertion must be a string");
  const parts = compact.split(".");
  if (parts.length !== 3 || parts[0] !== ASSERTION_PREFIX) fail("MALFORMED_ASSERTION", "invalid assertion envelope");
  const payloadBytes = b64urlDecode(parts[1]);
  const signature = b64urlDecode(parts[2]);
  if (signature.length !== 64) fail("MALFORMED_ASSERTION", "Ed25519 signatures must be 64 bytes");
  let parsed: unknown;
  try {
    parsed = JSON.parse(payloadBytes.toString("utf8"));
  } catch (error) {
    fail("MALFORMED_ASSERTION", `invalid payload JSON: ${String(error)}`);
  }
  const payload = validatePayload(parsed);
  if (!canonicalJsonBytes(payload as unknown as JsonValue).equals(payloadBytes)) fail("MALFORMED_ASSERTION", "payload is not canonical JSON");
  return { payload, payloadSegment: parts[1], signature, compact };
}

export function verifySignature(parsed: ParsedAssertion, publicKey: import("node:crypto").KeyObject): void {
  const input = Buffer.from(`${ASSERTION_PREFIX}.${parsed.payloadSegment}`, "ascii");
  if (!verify(null, input, publicKey, parsed.signature)) fail("BAD_SIGNATURE", "assertion signature verification failed");
}

File: python/agentctl/protocol.py
"""Language-neutral Agent Action Assertion v1 primitives.

The protocol signs the compact payload bytes, not a reinterpreted semantic
request. The transmitted body is hashed separately by the caller.
"""

from __future__ import annotations

import base64
import hashlib
import json
import re
import time
import uuid
from dataclasses import dataclass
from typing import Any, Mapping
from urllib.parse import quote, unquote_plus, urlsplit

from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)

PROTOCOL_VERSION = "agent-action-assertion/v1"
ASSERTION_PREFIX = "agentctl-aav1"
MAX_TTL_SECONDS = 300
_HEX_ESCAPE = re.compile(r"%([0-9A-Fa-f]{2})")
_JTI = re.compile(r"^[A-Za-z0-9._:-]{8,256}$")
_SHA256 = re.compile(r"^[0-9a-f]{64}$")
_METHOD = re.compile(r"^[A-Z]+$")
_CONTROL = re.compile(r"[\x00-\x20\x7f]")
_CONTROL_ONLY = re.compile(r"[\x00-\x1f\x7f]")


class AssertionErrorCode(ValueError):
    """Structured protocol failure with a stable machine-readable code."""

    def __init__(self, code: str, message: str):
        super().__init__(message)
        self.code = code
        self.message = message


def _fail(code: str, message: str) -> None:
    raise AssertionErrorCode(code, message)


def _validate_json_value(value: Any) -> None:
    if isinstance(value, bool) or value is None or isinstance(value, str):
        return
    if isinstance(value, int):
        return
    if isinstance(value, float):
        _fail("INVALID_JSON_NUMBER", "floating-point values are not allowed")
    if isinstance(value, list):
        for item in value:
            _validate_json_value(item)
        return
    if isinstance(value, dict):
        for key, item in value.items():
            if not isinstance(key, str):
                _fail("INVALID_JSON_KEY", "object keys must be strings")
            _validate_json_value(item)
        return
    _fail("INVALID_JSON_VALUE", f"unsupported JSON value type: {type(value).__name__}")


def canonical_json_bytes(value: Mapping[str, Any] | list[Any]) -> bytes:
    """Serialize values using the protocol's deterministic JSON rules."""

    _validate_json_value(value)
    try:
        text = json.dumps(
            value,
            ensure_ascii=False,
            allow_nan=False,
            separators=(",", ":"),
            sort_keys=True,
        )
    except (TypeError, ValueError) as exc:
        _fail("INVALID_CANONICAL_JSON", str(exc))
    return text.encode("utf-8")


def sha256_hex(value: bytes) -> str:
    return hashlib.sha256(value).hexdigest()


def normalize_content_type(value: str | None) -> str:
    if value is None:
        return ""
    if not isinstance(value, str):
        _fail("INVALID_CONTENT_TYPE", "content type must be a string")
    normalized = value.strip().lower()
    if _CONTROL_ONLY.search(normalized):
        _fail("INVALID_CONTENT_TYPE", "content type contains control characters")
    return normalized


def _validate_percent_escapes(value: str, code: str) -> None:
    position = 0
    while position < len(value):
        if value[position] == "%":
            if position + 2 >= len(value) or not re.fullmatch(
                r"[0-9A-Fa-f]{2}", value[position + 1 : position + 3]
            ):
                _fail(code, "invalid percent escape")
            position += 3
            continue
        position += 1


def _normalize_path(path: str) -> str:
    if path == "":
        path = "/"
    if not path.startswith("/"):
        _fail("INVALID_PATH", "request path must start with '/'")
    if _CONTROL.search(path):
        _fail("INVALID_PATH", "request path contains control characters or spaces")
    _validate_percent_escapes(path, "INVALID_PATH")

    output: list[str] = []
    position = 0
    while position < len(path):
        match = _HEX_ESCAPE.match(path, position)
        if match:
            byte = int(match.group(1), 16)
            char = chr(byte)
            if (char.isalnum() and byte < 128) or char in "-._~":
                output.append(char)
            else:
                output.append(f"%{byte:02X}")
            position = match.end()
            continue
        output.append(path[position])
        position += 1
    normalized = "".join(output)
    if any(segment in {".", ".."} for segment in normalized.split("/")):
        _fail("INVALID_PATH", "dot segments are not allowed")
    return normalized


def _canonical_query(query: str) -> str:
    if query == "":
        return ""
    _validate_percent_escapes(query, "INVALID_QUERY")
    pairs: list[tuple[str, str]] = []
    for raw_pair in query.split("&"):
        name, separator, value = raw_pair.partition("=")
        if not separator:
            value = ""
        try:
            decoded_name = unquote_plus(name, encoding="utf-8", errors="strict")
            decoded_value = unquote_plus(value, encoding="utf-8", errors="strict")
        except UnicodeDecodeError as exc:
            _fail("INVALID_QUERY", f"query is not valid UTF-8: {exc}")
        pairs.append((decoded_name, decoded_value))
    pairs.sort(key=lambda pair: (pair[0], pair[1]))
    encoded = [
        f"{quote(name, safe='-._~')}={quote(value, safe='-._~')}"
        for name, value in pairs
    ]
    return "&".join(encoded)


def canonical_request_target(target: str) -> str:
    """Return the canonical path plus sorted, percent-encoded query string."""

    if not isinstance(target, str) or not target:
        _fail("INVALID_PATH", "request target must be a non-empty string")
    try:
        parsed = urlsplit(target)
        if parsed.netloc:
            _ = parsed.port
    except ValueError as exc:
        _fail("INVALID_PATH", f"invalid request target: {exc}")
    if "#" in target or parsed.fragment:
        _fail("INVALID_PATH", "URL fragments are not allowed")
    if parsed.scheme and parsed.scheme.lower() not in {"http", "https"}:
        _fail("INVALID_PATH", "only HTTP and HTTPS URLs are supported")
    if parsed.username is not None or parsed.password is not None:
        _fail("INVALID_PATH", "URL userinfo is not allowed")
    path = _normalize_path(parsed.path)
    query = _canonical_query(parsed.query)
    return f"{path}?{query}" if query else path


def _require_string(payload: Mapping[str, Any], field: str, *, pattern: re.Pattern[str] | None = None) -> str:
    value = payload.get(field)
    if not isinstance(value, str) or not value.strip():
        _fail("MALFORMED_ASSERTION", f"{field} must be a non-empty string")
    if value != value.strip():
        _fail("MALFORMED_ASSERTION", f"{field} must be trimmed")
    if _CONTROL.search(value):
        _fail("MALFORMED_ASSERTION", f"{field} contains control characters")
    if pattern and not pattern.fullmatch(value):
        _fail("MALFORMED_ASSERTION", f"{field} has an invalid format")
    return value


def _require_int(payload: Mapping[str, Any], field: str) -> int:
    value = payload.get(field)
    if isinstance(value, bool) or not isinstance(value, int) or value < 0:
        _fail("MALFORMED_ASSERTION", f"{field} must be a non-negative integer")
    return value


def validate_payload(payload: Mapping[str, Any]) -> dict[str, Any]:
    allowed = {
        "version", "iss", "sub", "aud", "environment", "scope", "iat", "nbf", "exp",
        "jti", "kid", "http_method", "canonical_path", "content_type", "body_sha256",
        "request_id", "principal_epoch", "key_epoch", "resource", "project",
    }
    unknown = set(payload) - allowed
    if unknown:
        _fail("MALFORMED_ASSERTION", f"unknown assertion fields: {sorted(unknown)}")
    if payload.get("version") != PROTOCOL_VERSION:
        _fail("WRONG_VERSION", "unsupported assertion version")
    for field in ("iss", "sub", "aud", "environment", "scope", "kid", "request_id"):
        _require_string(payload, field)
    _require_string(payload, "jti", pattern=_JTI)
    _require_string(payload, "http_method", pattern=_METHOD)
    canonical_path = _require_string(payload, "canonical_path")
    if canonical_path != canonical_request_target(canonical_path):
        _fail("MALFORMED_ASSERTION", "canonical_path is not canonical")
    content_type = payload.get("content_type")
    if not isinstance(content_type, str) or content_type != normalize_content_type(content_type):
        _fail("MALFORMED_ASSERTION", "content_type is not normalized")
    body_sha256 = _require_string(payload, "body_sha256")
    if not _SHA256.fullmatch(body_sha256):
        _fail("MALFORMED_ASSERTION", "body_sha256 must be lowercase SHA-256 hex")
    for field in ("iat", "nbf", "exp", "principal_epoch", "key_epoch"):
        _require_int(payload, field)
    if payload["iss"] != payload["sub"]:
        _fail("ISSUER_SUBJECT_MISMATCH", "V1 does not support delegated subjects")
    if payload["iat"] > payload["nbf"] or payload["exp"] <= payload["nbf"]:
        _fail("INVALID_TIME_WINDOW", "iat <= nbf < exp is required")
    if payload["exp"] - payload["iat"] > MAX_TTL_SECONDS:
        _fail("TTL_EXCEEDED", "assertion TTL exceeds the V1 maximum")
    for field in ("resource", "project"):
        if field in payload:
            _require_string(payload, field)
    return dict(payload)


def _b64_encode(value: bytes) -> str:
    return base64.urlsafe_b64encode(value).decode("ascii").rstrip("=")


def _b64_decode(value: str) -> bytes:
    if not value or not re.fullmatch(r"[A-Za-z0-9_-]+", value):
        _fail("MALFORMED_ASSERTION", "invalid base64url segment")
    try:
        return base64.urlsafe_b64decode(value + "=" * (-len(value) % 4))
    except (ValueError, base64.binascii.Error) as exc:
        _fail("MALFORMED_ASSERTION", f"invalid base64url segment: {exc}")


@dataclass(frozen=True)
class ParsedAssertion:
    payload: dict[str, Any]
    payload_segment: str
    signature: bytes
    compact: str


def build_assertion(
    *,
    principal_id: str,
    audience: str,
    environment: str,
    scope: str,
    http_method: str,
    target: str,
    body: bytes,
    key_id: str,
    principal_epoch: int,
    key_epoch: int,
    private_key: Ed25519PrivateKey,
    content_type: str | None = None,
    resource: str | None = None,
    project: str | None = None,
    request_id: str | None = None,
    jti: str | None = None,
    now: int | None = None,
    ttl_seconds: int = 300,
) -> str:
    created = int(time.time()) if now is None else now
    if isinstance(created, bool) or not isinstance(created, int) or created < 0:
        _fail("INVALID_TIME_WINDOW", "now must be a non-negative integer")
    if isinstance(ttl_seconds, bool) or not isinstance(ttl_seconds, int):
        _fail("TTL_EXCEEDED", "ttl_seconds must be an integer")
    if ttl_seconds <= 0 or ttl_seconds > MAX_TTL_SECONDS:
        _fail("TTL_EXCEEDED", "ttl_seconds must be between 1 and 300")
    if not isinstance(http_method, str):
        _fail("INVALID_METHOD", "HTTP method must be a string")
    payload: dict[str, Any] = {
        "aud": _require_string({"value": audience}, "value"),
        "body_sha256": sha256_hex(body),
        "canonical_path": canonical_request_target(target),
        "content_type": normalize_content_type(content_type),
        "environment": _require_string({"value": environment}, "value"),
        "exp": created + ttl_seconds,
        "iat": created,
        "iss": _require_string({"value": principal_id}, "value"),
        "jti": str(uuid.uuid4()) if jti is None else jti,
        "kid": _require_string({"value": key_id}, "value"),
        "nbf": created,
        "http_method": http_method.upper(),
        "principal_epoch": principal_epoch,
        "request_id": str(uuid.uuid4()) if request_id is None else request_id,
        "scope": _require_string({"value": scope}, "value"),
        "sub": _require_string({"value": principal_id}, "value"),
        "version": PROTOCOL_VERSION,
        "key_epoch": key_epoch,
    }
    if resource is not None:
        payload["resource"] = resource
    if project is not None:
        payload["project"] = project
    payload = validate_payload(payload)
    payload_segment = _b64_encode(canonical_json_bytes(payload))
    signing_input = f"{ASSERTION_PREFIX}.{payload_segment}".encode("ascii")
    signature = private_key.sign(signing_input)
    return f"{ASSERTION_PREFIX}.{payload_segment}.{_b64_encode(signature)}"


def parse_assertion(compact: str) -> ParsedAssertion:
    if not isinstance(compact, str):
        _fail("MALFORMED_ASSERTION", "assertion must be a string")
    parts = compact.split(".")
    if len(parts) != 3 or parts[0] != ASSERTION_PREFIX:
        _fail("MALFORMED_ASSERTION", "invalid assertion envelope")
    payload_bytes = _b64_decode(parts[1])
    signature = _b64_decode(parts[2])
    if len(signature) != 64:
        _fail("MALFORMED_ASSERTION", "Ed25519 signatures must be 64 bytes")
    try:
        payload = json.loads(payload_bytes.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError) as exc:
        _fail("MALFORMED_ASSERTION", f"invalid payload JSON: {exc}")
    if not isinstance(payload, dict):
        _fail("MALFORMED_ASSERTION", "payload must be a JSON object")
    validate_payload(payload)
    if canonical_json_bytes(payload) != payload_bytes:
        _fail("MALFORMED_ASSERTION", "payload is not canonical JSON")
    return ParsedAssertion(
        payload=dict(payload),
        payload_segment=parts[1],
        signature=signature,
        compact=compact,
    )


def serialize_assertion(parsed: ParsedAssertion) -> str:
    return parsed.compact


def verify_signature(parsed: ParsedAssertion, public_key: Ed25519PublicKey) -> None:
    signing_input = f"{ASSERTION_PREFIX}.{parsed.payload_segment}".encode("ascii")
    try:
        public_key.verify(parsed.signature, signing_input)
    except Exception as exc:  # cryptography uses InvalidSignature without a common base.
        _fail("BAD_SIGNATURE", f"assertion signature verification failed: {exc}")
